### PR TITLE
TestEngine: Fix some snapshots problems

### PR DESCRIPTION
**Pull request by shargon**
- The snapshot where is readed the smart contract the first time was wrong.
- When a mock is called, the engine's storage is the current in use, otherwise, you haven't got access to the current data.
- Update readme with source coverage

### CODE REVIEW

File: src/Neo.SmartContract.Testing/TestingApplicationEngine.cs [hunk 1]
```diff
@@ -208,8 +209,30 @@ protected override void OnSysCall(InteropDescriptor descriptor)
...
 
+                    EngineStorage backup = Engine.Storage;
                     var hasReturnValue = customMock.Method.ReturnType != typeof(void);
-                    var returnValue = customMock.Method.Invoke(customMock.Contract, parameters);
+                    object returnValue;
+
+                    try
+                    {
+                        // We need to switch the Engine's snapshot in case
+                        // that a mock want to query the storage, it's not commited
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/TestingApplicationEngine.cs:219]:
```suggestion
                        // that a mock want to query the storage, it's not committed
```

File: tests/Neo.SmartContract.Template.UnitTests/templates/neocontractnep17/CoverageContractTests.cs [hunk 2]
```diff
@@ -15,24 +15,35 @@ public class CoverageContractTests
...
         public static void EnsureCoverage()
         {
-            // Join here all of your Coverage sources
+            // Join here all of your coverage sources
 
             var coverage = Nep17ContractTests.Coverage;
             coverage?.Join(OwnerContractTests.Coverage);
 
-            // Ensure that the coverage is more than X% at the end of the tests
+            // Dump coverate to console
```
Comment (shargon) [inline tests/Neo.SmartContract.Template.UnitTests/templates/neocontractnep17/CoverageContractTests.cs:23]:
```suggestion
            // Dump coverage to console
```

File: src/Neo.SmartContract.Testing/README.md [hunk 3]
```diff
@@ -416,7 +416,53 @@ public class CoverageContractTests
...
+
+    [AssemblyCleanup]
+    public static void EnsureCoverage()
+    {
+        // Join here all of your coverage sources
+
+        var coverage = Nep17ContractTests.Coverage;
+        coverage?.Join(OwnerContractTests.Coverage);
+
+        // Dump coverate to console
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/README.md:438]:
```suggestion
        // Dump coverage to console
```

File: src/Neo.SmartContract.Testing/README.md [hunk 4]
```diff
@@ -242,7 +242,7 @@ Assert.AreEqual(123, neo.BalanceOf(engine.ValidatorsAddress));
 
 using (ScriptBuilder script = new())
 {
-    script.EmitDynamicCall(neo.Hash, nameof(neo.BalanceOf), engine.ValidatorsAddress);
+    script.EmitDynamicCall(neo.Hash, "balanceOf", engine.ValidatorsAddress);
```
Comment (vncoelho) [inline src/Neo.SmartContract.Testing/README.md:245]: what is the difference?

Comment (shargon) [reply src/Neo.SmartContract.Testing/README.md:245]: the method start in lowercase, in neo the method start like this

File: src/Neo.SmartContract.Testing/README.md [hunk 5]
```diff
@@ -416,7 +416,53 @@ public class CoverageContractTests
 }
 ```
 
-Keep in mind that the coverage is at the instruction level. 
+Keep in mind that the coverage is at the instruction level, but you can also get the project's coverage on the source code using the debug file (`*.nefdbgnfo`) generated by `Neo.Compiler.CSharp`. To do this, you need to compile the project with the `-d` or `--debug` argument, and set up a unit test like the following:
```
Comment (vncoelho) [inline src/Neo.SmartContract.Testing/README.md:419]: what do you mean?

That the coverage here is at the instruction of the NVM, right?
What do you mean by project's coverage?

Comment (shargon) [reply src/Neo.SmartContract.Testing/README.md:419]: the source code, now you can get both of them

File: src/Neo.SmartContract.Testing/README.md (hunk 4 above)
Comment (vncoelho) [reply src/Neo.SmartContract.Testing/README.md:245]: `nameof(neo.BalanceOf)` was wrong?

Comment (shargon) [reply src/Neo.SmartContract.Testing/README.md:245]: Yes, this method doesn't exist

**Review by vncoelho — APPROVED** [review]

**Pull request merged**